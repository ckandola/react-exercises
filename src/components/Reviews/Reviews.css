/*
=============== 
Variables
===============
*/

:root {
    --clr-primary-1: hsl(205, 86%, 17%);
    --clr-primary-5: hsl(205, 78%, 60%);
    --clr-primary-7: hsl(205, 90%, 76%);
    --clr-primary-10: hsl(205, 100%, 96%);
    --clr-white: #fff;
    --transition: all 0.3s linear;
    --radius: 0.25rem;
    --light-shadow: 0 5px 15px rgba(0, 0, 0, 0.1);
    --dark-shadow: 0 5px 15px rgba(0, 0, 0, 0.2);
    --max-width: 1170px;
    --fixed-width: 620px;
  }
  
  /* section */
  .reviews-section {
    display: contents;
    width: 90vw;
    margin: 0 auto;
    max-width: var(--max-width);
  }
  
  /*
  =============== 
  Reviews
  ===============
  */
  .reviews-main {
    min-height: 100vh;
    display: grid;
    place-items: center;
  }
  .reviews-title {
    text-align: center;
    margin-bottom: 2rem;
  }
  .reviews-underline {
    height: 0.25rem;
    width: 5rem;
    background: var(--clr-primary-5);
    margin-left: auto;
    margin-right: auto;
  }
  .reviews-container {
    width: 80vw;
    max-width: var(--fixed-width);
  }
  .review {
    background: var(--clr-white);
    padding: 1.5rem 2rem;
    border-radius: var(--radius);
    box-shadow: var(--light-shadow);
    transition: var(--transition);
    text-align: center;
  }
  .review:hover {
    box-shadow: var(--dark-shadow);
  }
  .reviews-img-container {
    position: relative;
    width: 150px;
    height: 150px;
    border-radius: 50%;
    margin: 0 auto;
    margin-bottom: 1.5rem;
  }
  .reviews-person-img {
    width: 100%;
    display: block;
    height: 100%;
    object-fit: cover;
    border-radius: 50%;
    position: relative;
  }
  .reviews-quote-icon {
    position: absolute;
    top: 0;
    left: 0;
    width: 2.5rem;
    height: 2.5rem;
    display: grid;
    place-items: center;
    border-radius: 50%;
    transform: translateY(25%);
    background: var(--clr-primary-5);
    color: var(--clr-white);
  }
  .reviews-img-container::before {
    content: '';
    width: 100%;
    height: 100%;
    background: var(--clr-primary-5);
    position: absolute;
    top: -0.25rem;
    right: -0.5rem;
    border-radius: 50%;
  }
  .reviews-author {
    margin-bottom: 0.25rem;
  }
  .reviews-job {
    margin-bottom: 0.5rem;
    text-transform: uppercase;
    color: var(--clr-primary-5);
    font-size: 0.85rem;
  }
  .reviews-info {
    margin-bottom: 0.75rem;
  }
  .reviews-prev-btn,
  .reviews-next-btn {
    color: var(--clr-primary-7);
    font-size: 1.25rem;
    background: transparent;
    border-color: transparent;
    margin: 0 0.5rem;
    transition: var(--transition);
    cursor: pointer;
  }
  .reviews-prev-btn:hover,
  .reviews-next-btn:hover {
    color: var(--clr-primary-5);
  }
  .reviews-random-btn {
    margin-top: 0.5rem;
    background: var(--clr-primary-10);
    color: var(--clr-primary-5);
    padding: 0.25rem 0.5rem;
    text-transform: capitalize;
    border-radius: var(--radius);
    transition: var(--transition);
    border-color: transparent;
    cursor: pointer;
  }
  .reviews-random-btn:hover {
    background: var(--clr-primary-5);
    color: var(--clr-primary-1);
  }
  