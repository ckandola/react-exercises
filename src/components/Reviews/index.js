// from https://www.youtube.com/watch?v=a_7Z7C_JCyo

import React, {useState} from 'react';
import './Reviews.css';
import reviews from './data';
import { FaChevronLeft, FaChevronRight, FaQuoteRight} from 'react-icons/fa';

const Reviews = () => {
    const [reviewIx, setReviewIx] = useState(0);
    let review = reviews[reviewIx];

    const goToNextReview = () => {
        setReview(() => reviewIx + 1 === reviews.length ? 0 : reviewIx + 1);
    }
    
    const goToLastReview = () => {
        setReview(() => reviewIx - 1 < 0 ? reviews.length - 1 : reviewIx - 1)
    }

    const goToRandomReview = () => {
        setReview(() => {
            let temp = reviewIx;
            while (temp === reviewIx) {
                temp = Math.floor(Math.random() * reviews.length);
            }
            return temp;
        });
    }

    const setReview = setFunction => {
        const tempIx = setFunction();
        review = reviews[tempIx];
        setReviewIx(tempIx);
    }

    return (
        <main className="reviews-main">
            <section className="reviews-section">
                <div className="reviews-title">
                    <h2>Our Reviews</h2>
                    <div className="reviews-underline"></div>
                </div>
                <div className="reviews-container review">
                    <div className="reviews-img-container">
                        <img className="reviews-person-img" src={review.image} alt={review.name} />
                        <div className="reviews-quote-icon"><FaQuoteRight/></div>
                    </div>
                    <div className="reviews-author"><h4>{review.name}</h4></div>
                    <div className="reviews-job">{review.job}</div>
                    <div className="reviews-info">{review.text}</div>
                    <span>
                        <button className="reviews-prev-btn" onClick={goToLastReview}><FaChevronLeft /></button>
                        <button className="reviews-next-btn" onClick={goToNextReview}><FaChevronRight /></button>
                    </span>
                    <div>
                        <button className="reviews-random-btn" onClick={goToRandomReview}>Random Review</button>
                    </div>
                </div>
            </section>
        </main>
    );
}

export default Reviews;
