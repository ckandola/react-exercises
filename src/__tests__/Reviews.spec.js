import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Reviews from '../components/Reviews';

describe('Reviews Tests', () => {
    it('matches the snapshot', () => {
        const wrapper = shallow(<Reviews />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('advances to next review (anna johnson) when right arrow is clicked (from susan smith)', () => {
        const wrapper = shallow(<Reviews />);
        expect(wrapper.find('.reviews-person-img').getElement().props.alt).toEqual('susan smith');
        wrapper.find('.reviews-next-btn').simulate('click');
        expect(wrapper.find('.reviews-person-img').getElement().props.alt).toEqual('anna johnson');
    });
    
    it('advances to previous review when left arrow is clicked', () => {
        const wrapper = shallow(<Reviews />);
        expect(wrapper.find('.reviews-person-img').getElement().props.alt).toEqual('susan smith');
        wrapper.find('.reviews-next-btn').simulate('click');
        wrapper.find('.reviews-next-btn').simulate('click');
        wrapper.find('.reviews-prev-btn').simulate('click');
        expect(wrapper.find('.reviews-person-img').getElement().props.alt).toEqual('anna johnson');
    });
    
    it('advances to first review when right arrow is clicked after reaching end', () => {
        const wrapper = shallow(<Reviews />);
        expect(wrapper.find('.reviews-person-img').getElement().props.alt).toEqual('susan smith');
        wrapper.find('.reviews-next-btn').simulate('click');    // 2
        wrapper.find('.reviews-next-btn').simulate('click');    // 3
        wrapper.find('.reviews-next-btn').simulate('click');    // 4
        wrapper.find('.reviews-next-btn').simulate('click');    // 1
        expect(wrapper.find('.reviews-person-img').getElement().props.alt).toEqual('susan smith');
    });
    
    it('advances to last review (bill anderson) when left arrow is clicked at the beginning', () => {
        const wrapper = shallow(<Reviews />);
        expect(wrapper.find('.reviews-person-img').getElement().props.alt).toEqual('susan smith');
        wrapper.find('.reviews-prev-btn').simulate('click');
        expect(wrapper.find('.reviews-person-img').getElement().props.alt).toEqual('bill anderson');
    });
    
    it('changes to a different review when Random Review is clicked', () => {
        const wrapper = shallow(<Reviews />);
        expect(wrapper.find('.reviews-person-img').getElement().props.alt).toEqual('susan smith');
        wrapper.find('.reviews-random-btn').simulate('click');
        expect(wrapper.find('.reviews-person-img').getElement().props.alt).not.toEqual('susan smith');
    });
});
