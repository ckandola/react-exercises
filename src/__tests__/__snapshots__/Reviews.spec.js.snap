// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews Tests matches the snapshot 1`] = `
<main
  className="reviews-main"
>
  <section
    className="reviews-section"
  >
    <div
      className="reviews-title"
    >
      <h2>
        Our Reviews
      </h2>
      <div
        className="reviews-underline"
      />
    </div>
    <div
      className="reviews-container review"
    >
      <div
        className="reviews-img-container"
      >
        <img
          alt="susan smith"
          className="reviews-person-img"
          src="https://res.cloudinary.com/diqqf3eq2/image/upload/v1586883334/person-1_rfzshl.jpg"
        />
        <div
          className="reviews-quote-icon"
        >
          <FaQuoteRight />
        </div>
      </div>
      <div
        className="reviews-author"
      >
        <h4>
          susan smith
        </h4>
      </div>
      <div
        className="reviews-job"
      >
        web developer
      </div>
      <div
        className="reviews-info"
      >
        I'm baby meggings twee health goth +1. Bicycle rights tumeric chartreuse before they sold out chambray pop-up. Shaman humblebrag pickled coloring book salvia hoodie, cold-pressed four dollar toast everyday carry
      </div>
      <span>
        <button
          className="reviews-prev-btn"
          onClick={[Function]}
        >
          <FaChevronLeft />
        </button>
        <button
          className="reviews-next-btn"
          onClick={[Function]}
        >
          <FaChevronRight />
        </button>
      </span>
      <div>
        <button
          className="reviews-random-btn"
          onClick={[Function]}
        >
          Random Review
        </button>
      </div>
    </div>
  </section>
</main>
`;
